Unittests get stuck silently if there are non-daemon threads in the test

```
[DEBUG/MainProcess] received (20, 1, 'startTest', 21, (<handler_tests.TestHandler object at 0x7fa0a01f12b0>,), {})
[DEBUG/MainProcess] calling 'startTest'.<xmlrunner.result._XMLTestResult run=21 errors=20 failures=0> (remote ID 1) with args=(<handler_tests.TestHandler object at 0x7fa0a01f12b0>,), kwargs={}
[DEBUG/MainProcess] sending (20, 24, '__str__', 21, (), {})
[DEBUG/unittest] received (20, 24, '__str__', 21, (), {})
[DEBUG/unittest] calling '__str__'.<handler_tests.TestHandler testMethod=test_unstructured_message> (remote ID 24) with args=(), kwargs={}
[DEBUG/unittest] sending (25, 'test_unstructured_message (handler_tests.TestHandler)')
[DEBUG/MainProcess] received (25, 'test_unstructured_message (handler_tests.TestHandler)')
[DEBUG/MainProcess] sending (20, 24, '__str__', 21, (), {})
[DEBUG/unittest] received (20, 24, '__str__', 21, (), {})
[DEBUG/unittest] calling '__str__'.<handler_tests.TestHandler testMethod=test_unstructured_message> (remote ID 24) with args=(), kwargs={}
[DEBUG/unittest] sending (25, 'test_unstructured_message (handler_tests.TestHandler)')
[DEBUG/MainProcess] received (25, 'test_unstructured_message (handler_tests.TestHandler)')
[DEBUG] starting test_unstructured_message (handler_tests.TestHandler)
[DEBUG/MainProcess] sending (20, 24, '_testMethodName', 22)
[DEBUG/unittest] received (20, 24, '_testMethodName', 22)
[DEBUG/unittest] sending (25, 'test_unstructured_message')
[DEBUG/MainProcess] received (25, 'test_unstructured_message')
[DEBUG/MainProcess] sending (25, None)
[DEBUG/unittest] received (25, None)
[DEBUG/unittest] sending (20, 1, 'addError', 21, (<handler_tests.TestHandler testMethod=test_unstructured_message>, (<class 'AttributeError'>, AttributeError("module 'fluent.sender' has no attribute 'FluentSender'"), <traceback object at 0x7fdaaf8d1b40>)), {})
[DEBUG/MainProcess] received (20, 1, 'addError', 21, (<handler_tests.TestHandler object at 0x7fa0a01f12b0>, (<class 'AttributeError'>, AttributeError("module 'fluent.sender' has no attribute 'FluentSender'"), <traceback object at 0x7fa0a1a50e80>)), {})
[DEBUG/MainProcess] calling 'addError'.<xmlrunner.result._XMLTestResult run=22 errors=20 failures=0> (remote ID 1) with args=(<handler_tests.TestHandler object at 0x7fa0a01f12b0>, (<class 'AttributeError'>, AttributeError("module 'fluent.sender' has no attribute 'FluentSender'"), <traceback object at 0x7fa0a1a50e80>)), kwargs={}
[DEBUG/MainProcess] sending (20, 24, 'shortDescription', 21, (), {})
[DEBUG/unittest] received (20, 24, 'shortDescription', 21, (), {})
[DEBUG/unittest] calling 'shortDescription'.<handler_tests.TestHandler testMethod=test_unstructured_message> (remote ID 24) with args=(), kwargs={}
[DEBUG/unittest] sending (25, None)
[DEBUG/MainProcess] received (25, None)
[DEBUG/MainProcess] sending (20, 24, '__str__', 21, (), {})
[DEBUG/unittest] received (20, 24, '__str__', 21, (), {})
[DEBUG/unittest] calling '__str__'.<handler_tests.TestHandler testMethod=test_unstructured_message> (remote ID 24) with args=(), kwargs={}
[DEBUG/unittest] sending (25, 'test_unstructured_message (handler_tests.TestHandler)')
[DEBUG/MainProcess] received (25, 'test_unstructured_message (handler_tests.TestHandler)')
[DEBUG/MainProcess] sending (20, 24, 'failureException', 22)
[DEBUG/unittest] received (20, 24, 'failureException', 22)
[DEBUG/unittest] sending (25, <class 'AssertionError'>)
[DEBUG/MainProcess] received (25, <class 'AssertionError'>)
[DEBUG/MainProcess] sending (20, 24, 'id', 21, (), {})
[DEBUG/unittest] received (20, 24, 'id', 21, (), {})
[DEBUG/unittest] calling 'id'.<handler_tests.TestHandler testMethod=test_unstructured_message> (remote ID 24) with args=(), kwargs={}
[DEBUG/unittest] sending (25, 'handler_tests.TestHandler.test_unstructured_message')
[DEBUG/MainProcess] received (25, 'handler_tests.TestHandler.test_unstructured_message')
[DEBUG/MainProcess] sending (20, 24, 'failureException', 22)
[DEBUG/unittest] received (20, 24, 'failureException', 22)
[DEBUG/unittest] sending (25, <class 'AssertionError'>)
[DEBUG/MainProcess] received (25, <class 'AssertionError'>)
[DEBUG/MainProcess] sending (25, None)
[DEBUG/unittest] received (25, None)
[DEBUG/unittest] sending (20, 1, 'stopTest', 21, (<handler_tests.TestHandler testMethod=test_unstructured_message>,), {})
[DEBUG/MainProcess] received (20, 1, 'stopTest', 21, (<handler_tests.TestHandler object at 0x7fa0a01f12b0>,), {})
[DEBUG/MainProcess] calling 'stopTest'.<xmlrunner.result._XMLTestResult run=22 errors=21 failures=0> (remote ID 1) with args=(<handler_tests.TestHandler object at 0x7fa0a01f12b0>,), kwargs={}
[DEBUG/MainProcess] sending (25, None)
[DEBUG/unittest] received (25, None)
[DEBUG/unittest] sending (20, 1, 'shouldStop', 22)
[DEBUG/MainProcess] received (20, 1, 'shouldStop', 22)
[DEBUG/MainProcess] sending (25, False)
[DEBUG/unittest] received (25, False)
[DEBUG/unittest] sending (20, 1, '_testRunEntered', 22)
[DEBUG/MainProcess] received (20, 1, '_testRunEntered', 22)
[DEBUG/MainProcess] sending (25, True)
[DEBUG/unittest] received (25, True)
[DEBUG/unittest] sending (20, 1, '_previousTestClass', 22)
[DEBUG/MainProcess] received (20, 1, '_previousTestClass', 22)
[DEBUG/MainProcess] sending (25, <class 'handler_tests.TestHandler'>)
[DEBUG/unittest] received (25, <class 'handler_tests.TestHandler'>)
[DEBUG/unittest] sending (20, 1, '_moduleSetUpFailed', 22)
[DEBUG/MainProcess] received (20, 1, '_moduleSetUpFailed', 22)
[DEBUG/MainProcess] sending (25, False)
[DEBUG/unittest] received (25, False)
[DEBUG/unittest] sending (20, 1, '_setupStdout', 21, (), {})
[DEBUG/MainProcess] received (20, 1, '_setupStdout', 21, (), {})
[DEBUG/MainProcess] calling '_setupStdout'.<xmlrunner.result._XMLTestResult run=22 errors=21 failures=0> (remote ID 1) with args=(), kwargs={}
[DEBUG/MainProcess] sending (25, None)
[DEBUG/unittest] received (25, None)
[DEBUG/unittest] sending (20, 1, '_restoreStdout', 21, (), {})
[DEBUG/MainProcess] received (20, 1, '_restoreStdout', 21, (), {})
[DEBUG/MainProcess] calling '_restoreStdout'.<xmlrunner.result._XMLTestResult run=22 errors=21 failures=0> (remote ID 1) with args=(), kwargs={}
[DEBUG/MainProcess] sending (25, None)
[DEBUG/unittest] received (25, None)
[DEBUG/unittest] sending (20, 1, '_previousTestClass', 22)
[DEBUG/MainProcess] received (20, 1, '_previousTestClass', 22)
[DEBUG/MainProcess] sending (25, <class 'handler_tests.TestHandler'>)
[DEBUG/unittest] received (25, <class 'handler_tests.TestHandler'>)
[DEBUG/unittest] sending (20, 1, '_moduleSetUpFailed', 22)
[DEBUG/MainProcess] received (20, 1, '_moduleSetUpFailed', 22)
[DEBUG/MainProcess] sending (25, False)
[DEBUG/unittest] received (25, False)
[DEBUG/unittest] sending (20, 1, '_setupStdout', 21, (), {})
[DEBUG/MainProcess] received (20, 1, '_setupStdout', 21, (), {})
[DEBUG/MainProcess] calling '_setupStdout'.<xmlrunner.result._XMLTestResult run=22 errors=21 failures=0> (remote ID 1) with args=(), kwargs={}
[DEBUG/MainProcess] sending (25, None)
[DEBUG/unittest] received (25, None)
[DEBUG/unittest] sending (20, 1, '_restoreStdout', 21, (), {})
[DEBUG/MainProcess] received (20, 1, '_restoreStdout', 21, (), {})
[DEBUG/MainProcess] calling '_restoreStdout'.<xmlrunner.result._XMLTestResult run=22 errors=21 failures=0> (remote ID 1) with args=(), kwargs={}
[DEBUG/MainProcess] sending (25, None)
[DEBUG/unittest] received (25, None)
[DEBUG/unittest] sending (20, 1, '_testRunEntered', 23, False)
[DEBUG/MainProcess] received (20, 1, '_testRunEntered', 23, False)
[DEBUG/MainProcess] sending (25, None)
[DEBUG/unittest] received (25, None)
[DEBUG/unittest] sending (25, <xmlrunner.result._XMLTestResult object at 0x7fdaafd7e6d0>)
[DEBUG/MainProcess] received (25, <xmlrunner.result._XMLTestResult run=22 errors=21 failures=0>)
[DEBUG/MainProcess] sending (0, None)
[DEBUG/unittest] received (0, None)
[DEBUG/unittest] sending (1, None)
[DEBUG/MainProcess] received (1, None)
[INFO/unittest] process shutting down
[DEBUG/unittest] running all "atexit" finalizers with priority >= 0
[DEBUG/unittest] running the remaining "atexit" finalizers
^CTraceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/arcivanov/.pyenv/versions/3.9.7/lib/python3.9/multiprocessing/spawn.py", line 116, in spawn_main
    exitcode = _main(fd, parent_sentinel)
  File "/home/arcivanov/.pyenv/versions/3.9.7/lib/python3.9/multiprocessing/spawn.py", line 129, in _main
    return self._bootstrap(parent_sentinel)
  File "/home/arcivanov/.pyenv/versions/3.9.7/lib/python3.9/multiprocessing/process.py", line 333, in _bootstrap
    threading._shutdown()
  File "/home/arcivanov/.pyenv/versions/3.9.7/lib/python3.9/threading.py", line 1448, in _shutdown
    lock.acquire()
KeyboardInterrupt
Error in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/home/arcivanov/.pyenv/versions/3.9.7/lib/python3.9/logging/__init__.py", line 2142, in shutdown
Traceback (most recent call last):
  File "/home/arcivanov/.pyenv/versions/3.9.7/envs/fluent-logger-3.9.7/lib/python3.9/site-packages/pybuilder/plugins/python/unittest_plugin.py", line 153, in execute_tests_matching
    proc.join()
  File "/home/arcivanov/.pyenv/versions/3.9.7/envs/fluent-logger-3.9.7/lib/python3.9/site-packages/pybuilder/remote/__init__.py", line 149, in join
    return self.proc.join(timeout)
  File "/home/arcivanov/.pyenv/versions/3.9.7/lib/python3.9/multiprocessing/process.py", line 149, in join
    res = self._popen.wait(timeout)
  File "/home/arcivanov/.pyenv/versions/3.9.7/lib/python3.9/multiprocessing/popen_fork.py", line 43, in wait
    return self.poll(os.WNOHANG if timeout == 0.0 else 0)
  File "/home/arcivanov/.pyenv/versions/3.9.7/lib/python3.9/multiprocessing/popen_fork.py", line 27, in poll
    pid, sts = os.waitpid(self.pid, flag)
KeyboardInterrupt

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/arcivanov/.pyenv/versions/3.9.7/envs/fluent-logger-3.9.7/lib/python3.9/site-packages/pybuilder/cli.py", line 475, in main
    summary = reactor.build(
  File "/home/arcivanov/.pyenv/versions/3.9.7/envs/fluent-logger-3.9.7/lib/python3.9/site-packages/pybuilder/reactor.py", line 235, in build
    execution_summary = self.build_execution_plan(tasks, execution_plan)
  File "/home/arcivanov/.pyenv/versions/3.9.7/envs/fluent-logger-3.9.7/lib/python3.9/site-packages/pybuilder/reactor.py", line 270, in build_execution_plan
    task_execution_summaries = self.execution_manager.execute_execution_plan(
  File "/home/arcivanov/.pyenv/versions/3.9.7/envs/fluent-logger-3.9.7/lib/python3.9/site-packages/pybuilder/execution.py", line 393, in execute_execution_plan
    summaries.append(self.execute_task(task, **kwargs))
  File "/home/arcivanov/.pyenv/versions/3.9.7/envs/fluent-logger-3.9.7/lib/python3.9/site-packages/pybuilder/execution.py", line 340, in execute_task
    task.execute(self.logger, kwargs, _executable=_executable)
  File "/home/arcivanov/.pyenv/versions/3.9.7/envs/fluent-logger-3.9.7/lib/python3.9/site-packages/pybuilder/execution.py", line 179, in execute
    executable.execute(argument_dict)
  File "/home/arcivanov/.pyenv/versions/3.9.7/envs/fluent-logger-3.9.7/lib/python3.9/site-packages/pybuilder/execution.py", line 95, in execute
    self.callable(*arguments)
  File "/home/arcivanov/.pyenv/versions/3.9.7/envs/fluent-logger-3.9.7/lib/python3.9/site-packages/pybuilder/plugins/python/unittest_plugin.py", line 65, in run_unit_tests
    run_tests(project, logger, reactor, "unittest", "unit tests")
  File "/home/arcivanov/.pyenv/versions/3.9.7/envs/fluent-logger-3.9.7/lib/python3.9/site-packages/pybuilder/plugins/python/unittest_plugin.py", line 90, in run_tests
    result, console_out = execute_tests_matching(
  File "/home/arcivanov/.pyenv/versions/3.9.7/envs/fluent-logger-3.9.7/lib/python3.9/site-packages/pybuilder/plugins/python/unittest_plugin.py", line 156, in execute_tests_matching
    proc.close()
  File "/home/arcivanov/.pyenv/versions/3.9.7/envs/fluent-logger-3.9.7/lib/python3.9/site-packages/pybuilder/remote/__init__.py", line 155, in close
    return self.proc.close()
  File "/home/arcivanov/.pyenv/versions/3.9.7/lib/python3.9/multiprocessing/process.py", line 181, in close
    raise ValueError("Cannot close a process while it is still running. "
ValueError: Cannot close a process while it is still running. You should first call join() or terminate().
------------------------------------------------------------
BUILD FAILED - ValueError: Cannot close a process while it is still running. You should first call join() or terminate(). (multiprocessing/process.py:181)
------------------------------------------------------------
Build finished at 2021-10-09 18:33:15
Build took 28 seconds (28592 ms)
    h.close()
  File "/home/arcivanov/Documents/src/arcivanov/fluent-logger/src/main/python/fluent/handler.py", line 244, in close
    self.sender.close()
  File "/home/arcivanov/Documents/src/arcivanov/fluent-logger/src/main/python/fluent/handler.py", line 207, in sender
    self._sender = self.getSenderInstance(
  File "/home/arcivanov/Documents/src/arcivanov/fluent-logger/src/main/python/fluent/handler.py", line 223, in getSenderInstance
    sender_class = self.getSenderClass()
  File "/home/arcivanov/Documents/src/arcivanov/fluent-logger/src/main/python/fluent/handler.py", line 202, in getSenderClass
    return sender.FluentSender
AttributeError: module 'fluent.sender' has no attribute 'FluentSender'
```

--- a/src/integrationtest/python/issue_822_tests.py
+++ b/src/integrationtest/python/issue_822_tests.py
@@ -1,0 +1,76 @@
+#   -*- coding: utf-8 -*-
+#
+#   This file is part of PyBuilder
+#
+#   Copyright 2011-2021 PyBuilder Team
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+import textwrap
+import unittest
+
+from itest_support import IntegrationTestSupport
+from pybuilder.errors import BuildFailedException
+
+
+class Issue822Test(IntegrationTestSupport):
+    def test(self):
+        self.write_build_file("""
+from pybuilder.core import use_plugin, init
+
+use_plugin("python.core")
+use_plugin("python.unittest")
+
+@init
+def init (project):
+    project.set_property("verbose", True)
+    project.set_property("remote_debug", 2)
+    project.set_property("remote_tracing", 1)
+        """)
+
+        self.create_directory("src/main/python")
+        self.create_directory("src/unittest/python")
+        self.write_file("src/main/python/code.py", textwrap.dedent(
+            """
+            import threading
+            import time
+            class TestThread(threading.Thread):
+                def __init__(self):
+                    super().__init__()
+
+                def run(self):
+                    time.sleep(7)
+
+            def run_code():
+                TestThread().start()
+
+            """))
+        self.write_file("src/unittest/python/code_tests.py", textwrap.dedent(
+            """
+            import unittest
+            import code
+
+            class CodeTests(unittest.TestCase):
+                def test_code(self):
+                    code.run_code()
+            """
+        ))
+        reactor = self.prepare_reactor()
+        with self.assertRaises(BuildFailedException) as raised_ex:
+            reactor.build("verify")
+
+        self.assertEqual(raised_ex.exception.message, "Unittest tool failed with exit code 1")
+
+
+if __name__ == "__main__":
+    unittest.main()

--- a/src/integrationtest/python/issue_822_with_remote_debug_tests.py
+++ b/src/integrationtest/python/issue_822_with_remote_debug_tests.py
@@ -1,0 +1,75 @@
+#   -*- coding: utf-8 -*-
+#
+#   This file is part of PyBuilder
+#
+#   Copyright 2011-2021 PyBuilder Team
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+import textwrap
+import unittest
+
+from itest_support import IntegrationTestSupport
+from pybuilder.errors import BuildFailedException
+
+
+class Issue822Test(IntegrationTestSupport):
+    def test(self):
+        self.write_build_file("""
+from pybuilder.core import use_plugin, init
+
+use_plugin("python.core")
+use_plugin("python.unittest")
+
+@init
+def init (project):
+    project.set_property("verbose", True)
+    project.set_property("remote_debug", 1)
+        """)
+
+        self.create_directory("src/main/python")
+        self.create_directory("src/unittest/python")
+        self.write_file("src/main/python/code.py", textwrap.dedent(
+            """
+            import threading
+            import time
+            class TestThread(threading.Thread):
+                def __init__(self):
+                    super().__init__()
+
+                def run(self):
+                    time.sleep(7)
+
+            def run_code():
+                TestThread().start()
+
+            """))
+        self.write_file("src/unittest/python/code_tests.py", textwrap.dedent(
+            """
+            import unittest
+            import code
+
+            class CodeTests(unittest.TestCase):
+                def test_code(self):
+                    code.run_code()
+            """
+        ))
+        reactor = self.prepare_reactor()
+        with self.assertRaises(BuildFailedException) as raised_ex:
+            reactor.build("verify")
+
+        self.assertEqual(raised_ex.exception.message, "Unittest tool failed with exit code 1")
+
+
+if __name__ == "__main__":
+    unittest.main()

--- a/src/integrationtest/python/issue_822_with_tracing_tests.py
+++ b/src/integrationtest/python/issue_822_with_tracing_tests.py
@@ -1,0 +1,75 @@
+#   -*- coding: utf-8 -*-
+#
+#   This file is part of PyBuilder
+#
+#   Copyright 2011-2021 PyBuilder Team
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+import textwrap
+import unittest
+
+from itest_support import IntegrationTestSupport
+from pybuilder.errors import BuildFailedException
+
+
+class Issue822Test(IntegrationTestSupport):
+    def test(self):
+        self.write_build_file("""
+from pybuilder.core import use_plugin, init
+
+use_plugin("python.core")
+use_plugin("python.unittest")
+
+@init
+def init (project):
+    project.set_property("verbose", True)
+    project.set_property("remote_tracing", 1)
+        """)
+
+        self.create_directory("src/main/python")
+        self.create_directory("src/unittest/python")
+        self.write_file("src/main/python/code.py", textwrap.dedent(
+            """
+            import threading
+            import time
+            class TestThread(threading.Thread):
+                def __init__(self):
+                    super().__init__()
+
+                def run(self):
+                    time.sleep(7)
+
+            def run_code():
+                TestThread().start()
+
+            """))
+        self.write_file("src/unittest/python/code_tests.py", textwrap.dedent(
+            """
+            import unittest
+            import code
+
+            class CodeTests(unittest.TestCase):
+                def test_code(self):
+                    code.run_code()
+            """
+        ))
+        reactor = self.prepare_reactor()
+        with self.assertRaises(BuildFailedException) as raised_ex:
+            reactor.build("verify")
+
+        self.assertEqual(raised_ex.exception.message, "Unittest tool failed with exit code 1")
+
+
+if __name__ == "__main__":
+    unittest.main()

--- a/src/main/python/pybuilder/plugins/core_plugin.py
+++ b/src/main/python/pybuilder/plugins/core_plugin.py
@@ -31,6 +31,7 @@ def init(project):
     project.set_property("dir_logs", jp("$dir_target", "logs"))
 
     project.set_property_if_unset("remote_debug", 0)
+    project.set_property_if_unset("remote_tracing", 0)
 
     project.write_report = partial(write_report, project)
 

--- a/src/main/python/pybuilder/plugins/python/remote_tools/__init__.py
+++ b/src/main/python/pybuilder/plugins/python/remote_tools/__init__.py
@@ -16,7 +16,6 @@
 #   See the License for the specific language governing permissions and
 #   limitations under the License.
 
-
 from pybuilder.remote import Process, PipeShutdownError, RemoteObjectPipe, logger, log_to_stderr
 
 __all__ = ["RemoteObjectPipe", "start_tool", "Tool", "PipeShutdownError", "logger"]
@@ -32,7 +31,7 @@ class Tool:
         pass
 
 
-def start_tool(pyenv, tools, group=None, name=None, logging=None):
+def start_tool(pyenv, tools, group=None, name=None, logging=None, tracing=None):
     """
     Starts a tool process
     """
@@ -42,7 +41,8 @@ def start_tool(pyenv, tools, group=None, name=None, logging=None):
         logger.setLevel(int(logging))
 
     pipe = RemoteObjectPipe.new_pipe()
-    proc = Process(pyenv, group=group, target=_instrumented_tool, name=name, args=(tools, pipe))
+    proc = Process(pyenv, group=group, name=name,
+                   target=_traced_tool if tracing else _instrumented_tool, args=(tools, pipe))
 
     try:
         proc.start()
@@ -51,6 +51,16 @@ def start_tool(pyenv, tools, group=None, name=None, logging=None):
 
     pipe.receive()  # Pickle protocol selection
     return proc, pipe
+
+
+def _traced_tool(tools, pipe):
+    import trace
+
+    def _print(*objects, sep=' ', end='', **kwargs):
+        logger.debug((sep.join(objects) + end).rstrip("\r\n"))
+
+    trace.print = _print
+    trace.Trace(count=0).runfunc(_instrumented_tool, tools, pipe)
 
 
 def _instrumented_tool(tools, pipe):
@@ -69,4 +79,28 @@ def _instrumented_tool(tools, pipe):
     except Exception as e:
         pipe.close(e)
     finally:
-        pipe.close()
+        try:
+            pipe.close()
+        finally:
+            import threading
+
+            main = threading.main_thread()
+            current = threading.current_thread()
+
+            if main != current:
+                logger.warn("current thread %s is not the main %s in the tool process", current, main)
+
+            blocked_threads = False
+            for t in threading.enumerate():
+                if not t.daemon and t != current:
+                    logger.warn("non-daemon thread %s is blocking the tool process shutdown", t)
+                    blocked_threads = True
+
+            if blocked_threads:
+                import os
+                import atexit
+
+                try:
+                    atexit._run_exitfuncs()
+                finally:
+                    os._exit(1)

--- a/src/main/python/pybuilder/plugins/python/remote_tools/unittest_tool.py
+++ b/src/main/python/pybuilder/plugins/python/remote_tools/unittest_tool.py
@@ -56,6 +56,6 @@ class UnitTestTool(Tool):
         pipe.hide("unittest_tests")
 
 
-def start_unittest_tool(pyenv, tools, sys_paths, test_modules, test_method_prefix, logging=0):
+def start_unittest_tool(pyenv, tools, sys_paths, test_modules, test_method_prefix, logging=0, tracing=0):
     tool = UnitTestTool(sys_paths, test_modules, test_method_prefix)
-    return start_tool(pyenv, tools + [tool], name="unittest", logging=logging)
+    return start_tool(pyenv, tools + [tool], name="unittest", logging=logging, tracing=tracing)

--- a/src/main/python/pybuilder/plugins/python/unittest_plugin.py
+++ b/src/main/python/pybuilder/plugins/python/unittest_plugin.py
@@ -91,7 +91,8 @@ def run_tests(project, logger, reactor, execution_prefix, execution_name):
             reactor.python_env_registry[project.get_property("unittest_python_env")],
             reactor.tools, runner_generator, logger, test_dir, module_glob, [test_dir, src_dir],
             test_method_prefix,
-            project.get_property("remote_debug"))
+            project.get_property("remote_debug"),
+            project.get_property("remote_tracing"))
 
         if result.testsRun == 0:
             logger.warn("No %s executed.", execution_name)
@@ -121,13 +122,13 @@ def run_tests(project, logger, reactor, execution_prefix, execution_name):
 
 
 def execute_tests(pyenv, tools, runner_generator, logger, test_source, suffix, sys_paths, test_method_prefix=None,
-                  remote_debug=0):
+                  remote_debug=0, remote_tracing=0):
     return execute_tests_matching(pyenv, tools, runner_generator, logger, test_source, "*{0}".format(suffix),
-                                  test_method_prefix, remote_debug=remote_debug)
+                                  test_method_prefix, remote_debug=remote_debug, remote_tracing=remote_tracing)
 
 
 def execute_tests_matching(pyenv, tools, runner_generator, logger, test_source, file_glob, sys_paths,
-                           test_method_prefix=None, remote_debug=0):
+                           test_method_prefix=None, remote_debug=0, remote_tracing=0):
     output_log_file = StringIO()
     try:
         test_modules = discover_modules_matching(test_source, file_glob)
@@ -135,9 +136,10 @@ def execute_tests_matching(pyenv, tools, runner_generator, logger, test_source, 
                                     logger,
                                     _create_runner(runner_generator, output_log_file))
 
+        exit_code = None
         try:
             proc, pipe = start_unittest_tool(pyenv, tools, sys_paths, test_modules, test_method_prefix,
-                                             logging=remote_debug)
+                                             logging=remote_debug, tracing=remote_tracing)
             try:
                 pipe.register_remote(runner)
                 pipe.register_remote_type(unittest.result.TestResult)
@@ -153,9 +155,15 @@ def execute_tests_matching(pyenv, tools, runner_generator, logger, test_source, 
                         proc.join()
                     finally:
                         try:
-                            proc.close()
-                        except AttributeError:
-                            pass
+                            exit_code = proc.exitcode
+                        finally:
+                            try:
+                                proc.close()
+                            except AttributeError:
+                                pass
+
+            if exit_code:
+                raise BuildFailedException("Unittest tool failed with exit code %s", exit_code)
 
             remote_closed_cause = pipe.remote_close_cause()
             if remote_closed_cause is not None:

--- a/src/unittest/python/plugins/python/unittest_plugin_tests.py
+++ b/src/unittest/python/plugins/python/unittest_plugin_tests.py
@@ -49,7 +49,9 @@ class ExecuteTestsTests(TestCase):
     def test_should_discover_modules_by_suffix(self, mock_discover_modules_matching, mock_unittest, runner, tool):
         pipe = Mock()
         pipe.remote_close_cause.return_value = None
-        tool.return_value = (Mock(), pipe)
+        proc = Mock()
+        proc.exitcode = 0
+        tool.return_value = (proc, pipe)
         execute_tests(Mock(), [], runner, self.mock_logger, "/path/to/test/sources", "_tests.py", ["a", "b"])
 
         mock_discover_modules_matching.assert_called_with("/path/to/test/sources", "*_tests.py")
@@ -61,7 +63,9 @@ class ExecuteTestsTests(TestCase):
     def test_should_discover_modules_by_glob(self, mock_discover_modules_matching, mock_unittest, runner, tool):
         pipe = Mock()
         pipe.remote_close_cause.return_value = None
-        tool.return_value = (Mock(), pipe)
+        proc = Mock()
+        proc.exitcode = 0
+        tool.return_value = (proc, pipe)
         execute_tests_matching(Mock(), [], runner, self.mock_logger, "/path/to/test/sources", "*_tests.py", ["a", "b"])
 
         mock_discover_modules_matching.assert_called_with("/path/to/test/sources", "*_tests.py")
@@ -73,7 +77,9 @@ class ExecuteTestsTests(TestCase):
     def test_should_return_actual_test_results(self, mock_discover_modules, mock_unittest, runner, tool):
         pipe = Mock()
         pipe.remote_close_cause.return_value = None
-        tool.return_value = (Mock(), pipe)
+        proc = Mock()
+        proc.exitcode = 0
+        tool.return_value = (proc, pipe)
         mock_tests = Mock()
         mock_unittest.defaultTestLoader.loadTestsFromNames.return_value = mock_tests
         runner.return_value.run.return_value = self.mock_result

